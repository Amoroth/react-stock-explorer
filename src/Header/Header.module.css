.container {
  align-items: center;
  background-color: #344955;
  box-shadow: 2px 4px 10px rgba(0,0,0,.2);
  display: flex;
  height: 62px;
  justify-content: space-between;
  position: fixed;
  top: 0;
  width: 100%;
}

.brand {
  color: #fff;
  font-size: 22px;
  margin: 0 15px;
}

.nav-links {
  height: 100%;
  margin: 0 15px;
}

.nav-links a {
  color: #fff;
  display: inline-block;
  height: 100%;
  line-height: 60px;
  margin-left: 15px;
  position: relative;
  text-decoration: none;
}

.nav-links a:hover {
  color: #ddd;
}

.nav-link-active {
  border-bottom: 5px solid #fff;
}
.nav-link-active:hover {
  border-bottom: 5px solid #ddd;
}

@media screen and (max-width: 620px) {
  .nav-links {
    visibility: hidden;
  }
}