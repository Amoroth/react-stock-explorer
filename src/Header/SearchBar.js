import React, { Component } from 'react'
import { createFilter } from 'react-search-input'
import { Link } from 'react-router-dom'

import Overlay from '../shared/Overlay'
import styles from './Header.module.css'

class SearchBar extends Component {
  constructor(props) {
    super(props)
    this.state = {
      input: '',
      symbols: [],
      searchResults: false,
      searchBoxInput: false,
    }
    this.inputRef = React.createRef();
  }

  componentDidMount() {
    fetch('https://api.iextrading.com/1.0/ref-data/symbols?filter=symbol,name').then((res) => {
      return res.json()
    }).then((json) => {
      this.setState({ symbols: json, searchBoxInput: window.innerWidth > 800 })
    })
    this.setState({ symbols: [
      {name: 'Alphabet Inc.', symbol: 'googl'},
      {name: 'Microsoft Corporation', symbol: 'msft'},
      {name: 'International Buissnes Machines Corporation', symbol: 'ibm'},
      {name: 'Twitter Inc.', symbol: 'twtr'}
    ],
    searchBoxInput: window.innerWidth > 800 })
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.input !== prevState.input) {
      return true
    }
    return false
  }

  handleChange = (e) => {
    this.setState({ input: e.target.value })
  }

  onClickFill = (term) => {
    this.setState({ input: term })
  }

  showSearchResults = () => {
    this.setState({ searchResults: true })
  }

  hideSearchResults = () => {
    setTimeout(() => this.setState({ searchResults: false, searchBoxInput: window.innerWidth > 800 }), 100)
  }

  showSearchBoxInput = () => {
    this.setState({ searchBoxInput: true }, () => {
      this.inputRef.current.focus()
    })
  }

  render() {
    const filteredCmps = this.state.symbols.filter(
      createFilter(this.state.input, ['name', 'symbol'])
    )

    return (
      <div style={{display: 'inline'}}>
        {this.state.searchBoxInput ? 
        <input
          className={styles['search-box']}
          onChange={this.handleChange}
          value={this.state.input}
          onFocus={this.showSearchResults}
          onBlur={this.hideSearchResults}
          ref={this.inputRef}
        /> : 
        <button onClick={this.showSearchBoxInput} className={styles['search-box-icon']}>
          <i className="material-icons">search</i>
        </button>}

        <div className={styles['search-box-results']}>
          {this.state.searchResults ? filteredCmps.slice(0, 4).map((cmp) => {
            return (
              <Link
                className={styles['search-box-result']}
                key={cmp.symbol}
                onClick={() => this.onClickFill(cmp.name)}
                to={`stock?cmp=${cmp.symbol}`}>
                {cmp.name}
                <span>{cmp.symbol}</span>
              </Link>
            )}) : null}
        </div>
        {this.state.searchResults && window.innerWidth < 800 ? <Overlay /> : null}
      </div>
    )
  }
}

export default SearchBar