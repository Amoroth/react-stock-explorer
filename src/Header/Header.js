import React from 'react'
import { withRouter, NavLink } from 'react-router-dom'

import SearchBar from './SearchBar'
import styles from './Header.module.css'

const header = (props) => {
  let expandButton = (
  <button className={styles['expand-button']}>
    <i className="material-icons">menu</i>
  </button>)

  const onBackClick = () => {
    if (props.location.pathname === '/company') {
      props.history.goBack()
    } else {
      props.history.push('/')
    }
  }

  if (props.location.pathname !== '/') {
    expandButton = (
      <button className={styles['expand-button']} onClick={onBackClick}>
        <i className="material-icons">arrow_back</i>
      </button>
    )
  }

  let navlinkStyle = [styles['nav-link'], styles['desktop-only']].join(' ')

  return (
    <nav className={styles['container']}>
      <div className={styles['brand-container']}>
        { expandButton }
        <h6 className={styles['brand']}>Stock Explorer!</h6>
      </div>
      <div className={[styles['nav-links']].join(' ')}>
        <SearchBar />
        <NavLink to="/" exact activeClassName={styles['nav-link-active']} className={navlinkStyle}>Stocks</NavLink>
        <NavLink to="/error" activeClassName={styles['nav-link-active']} className={navlinkStyle}>Cryptos</NavLink>
      </div>
    </nav>
  )
}

export default withRouter(header)