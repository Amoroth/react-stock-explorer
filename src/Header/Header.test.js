import React from 'react'
import { shallow } from 'enzyme'

import Header from './Header'

describe('<Header />', () => {
  let props = {}
  let wrapper

  describe('desktop', () => {
    beforeEach(() => {
      props = {
        location: { pathname: '/market' },
        history: { goBack: jest.fn(), push: jest.fn() },
        currencyChange: jest.fn(),
        currency: 'USD',
      }
      wrapper = shallow(<Header.WrappedComponent {...props} />)
    })

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('renders brand as a Link', () => {
      expect(wrapper.find('Link').exists()).toBe(true)
      expect(wrapper.find('Link').props().to).toBe('/market')
    })

    it('renders dropdown', () => {
      expect(wrapper.find('Dropdown').exists()).toBe(true)
    })

    it('changes currency', () => {
      wrapper.find('Dropdown').simulate('change', { value: 'EUR' })
      expect(props.currencyChange).toBeCalledWith('EUR')
    })
  })

  describe('mobile', () => {
    beforeEach(() => {
      props = {
        location: { pathname: '/market' },
        history: { goBack: jest.fn(), push: jest.fn() },
        currencyChange: jest.fn(),
        currency: 'USD',
      }
      window.innerWidth = 400
      wrapper = shallow(<Header.WrappedComponent {...props} />)
    })

    it('renders brand on mobile', () => {
      expect(wrapper.find('h6').text()).toBe('SE!')
    })

    it('goes back from stock page', () => {
      wrapper.setProps({ location: { pathname: '/stock' } })

      wrapper.find('.expand-button').simulate('click')
      expect(props.history.goBack).not.toBeCalled()
      expect(props.history.push).toBeCalledWith('/market')
    })

    it('goes back from company page', () => {
      wrapper.setProps({ location: { pathname: '/company' } })

      wrapper.find('.expand-button').simulate('click')
      expect(props.history.goBack).toBeCalled()
      expect(props.history.push).not.toBeCalled()
    })
  })
})
