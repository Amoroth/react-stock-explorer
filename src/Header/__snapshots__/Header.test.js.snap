// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> desktop renders correctly 1`] = `
<nav
  className="container"
>
  <div
    className="brand-container"
  >
    <Link
      className="brand"
      replace={false}
      to="/market"
    >
      Stock Explorer!
    </Link>
  </div>
  <div
    className="nav-links"
  >
    <withRouter(SearchBar) />
    <Dropdown
      baseClassName="Dropdown"
      className="dropdown"
      controlClassName="dropdown-control"
      onChange={[Function]}
      options={
        Array [
          "USD",
          "EUR",
          "PLN",
          "GBP",
          "CHF",
        ]
      }
      value="USD"
    />
  </div>
</nav>
`;
